packages/aws_common/lib/src/logging/aws_logger.dart

```diff
@@ -72,10 +87,38 @@ class AWSLogger implements Closeable {
     return AWSLogger('$namespace.$name');
   }
 
+  /// Returns a plugin of type [Plugin] registered to this
+  /// logger hierarchy or null.
+  Plugin? getPlugin<Plugin extends AWSLoggerPlugin>() {
+    var registeredPlugin = parent?.getPlugin<Plugin>();
+    return registeredPlugin ??= _subscriptions.keys
+        .firstWhereOrNull((element) => element is Plugin) as Plugin?;
+  }
+
   /// Registers an [AWSLoggerPlugin] to handle logs emitted by this logger
   /// instance.
-  void registerPlugin(AWSLoggerPlugin plugin) {
-    unregisterPlugin(plugin);
+  ///
+  /// Throws [StateError] if a plugin with same type is registered to this
+  /// logger hierarchy.
+  /// Unregisteres plugins with same type from all children in the hierarchy
+  /// to avoid duplicate logging.
+  void registerPlugin<T extends AWSLoggerPlugin>(
+    T plugin,
+  ) {
+    final registeredPlugin = getPlugin<T>();
+    if (registeredPlugin != null) {
+      throw StateError(
+        'A plugin with same type is already registered in the logger hierarchy.'
+        ' Unregister the existing plugin first before registering a new plugin.',
+      );
+    }
+    // Unregisters plugin for all childeren to avoid duplicate logging.
+    for (final element in children) {
```

Comment:
Why only two levels deep?

What was the rationale for having only one plugin per type per hierarchy? I had remembered something more like the following:

```dart
final logger = AWSLogger();
final childLogger = logger.createChild('child');

final plugin1 = LoggerPlugin();
final plugin2 = LoggerPlugin();

logger.registerPlugin(plugin1);
childLogger.registerPlugin(plugin2);

logger.getPlugin<LoggerPlugin>(); // plugin1
childLogger.getPlugin<LoggerPlugin>(); // plugin2

childLogger.unregisterPlugin(plugin2);
childLogger.getPlugin<LoggerPlugin>(); // plugin1
```

Reply:
ops, will fix the loop.

We can do what you suggested with these considerations:
- logger.getPlugin<T>() and logger.configure(PluginKey, (plugin) {...}) returns/configures a plugin which is closest to the logger in the logger hierarchy. this can be a source of confusion for developers/customers.  
- When registering a plugin we should check if same plugin is registered in the hierarchy and not allow registering it again otherwise the plugin records same logs multiple times which is not expected/wanted.
  
We decided to allow at most on plugin per type in a logger hierarchy as we think this is the common way of using logger and logger plugin. This way getPlugin<T>() and logger.configure(PluginKey, (plugin) {...}) behavior won't cause any confusion for developers/customers.
wdyt?

Reply:
Oh yeah. I think the concession we made was that the plugin would be responsible for tracking hierarchical settings so that the settings on the root logger could differ from the child settings. Basically, we want to ensure that `AmplifyLogger()` and `AmplifyLogger.category(Category.auth)` can maintain different settings even though they're in the same hierarchy. 

If that's still the plan, then I'm okay with keeping the 1 logger type per/hierarchy but you'll need to address the subclassing issues I mentioned below e.g. when one logger plugin inherits from another. These are technically different classes but will conflict due to the implementation (i.e. checking `is Plugin`).

I would also prefer if `registerPlugin` did not automatically de-register existing plugins anymore, but instead threw an `Error` and required an explicit call to `unregisterPlugin` first since the new implicit behavior is no longer intuitive.

Reply:
re `Basically, we want to ensure that AmplifyLogger() and AmplifyLogger.category(Category.auth) can maintain different settings even though they're in the same hierarchy.`:  this is correct. plugin is responsible to keep track of settings for hierarchical loggers. e.g. the CW logger plugin tracks the logLevel for each logger it should send to CW.

re `subclassing issues`: that should be fixed using `.runtimeType is Plugin`.

re `registerPlugin does not automatically de-register existing plugins`: agree will update the code.